EIPS/eip-5450.md

```diff
@@ -13,51 +13,108 @@ requires: 3540, 3670, 4200, 4750
+by verifying that no exceptional conditions can happen
+and preventing the execution and deployment of any invalid code.
+
+The operand stack validation provides several benefits:
+
+- removes run-time stack underflow check for all instructions,
+- removes run-time stack overflow check for all instruction except `CALLF`,
+- ensures that an execution terminates with one of the terminating instructions,
+- prevents the deployment of code with unreachable instructions, which discourages the use of code sections for data storage.
+
+It also has some disadvantages:
+
+- adds constraints to the code structure (similar to JVM, CPython bytecode, WebAssembly and others); however, these constraints can be lifted in a backward-compatible manner if they are shown to be user-unfriendly,
+- adds second validation pass; however, validation's computational and space complexity remains linear.
 
 ## Specification
 
 ### Code validation
 
-*Remark:* We rely on the notions of *data stack* and *type section* as defined by [EIP-4750](./eip-4750.md).
+*Remark:* We rely on the notions of *operand stack* (FIXME) and *type section* as defined by [EIP-4750](./eip-4750.md).
+
+Each code section is validated independently.
+
+#### Instructions validation
+
+In the first validation phase defined in [EIP-3670](./eip-3670.md) (and extended by [EIP-4200](./eip-4200.md) and [EIP-4750](./eip-4750.md)) instructions are inspected independently to check if their opcodes and immediate values are valid.
+
+#### Operand stack validation
+
+In the second validation phase control-flow analysis is performed on the code.
 
-Code section validation rules as defined by [EIP-3670](./eip-3670.md) (which has been extended by [EIP-4200](./eip-4200.md) and [EIP-4750](./eip-4750.md)) are extended again to include the instruction flow traversal procedure, where every possible code path is examined, and data stack height at each instruction is recorded. 
+*Operand stack height* here refers to the number of stack values accessible by this function, i.e. it does not take into account values of caller functions' frames (but does include this function's inputs). Note that validation procedure does not require actual data stack implementation, but only to keep track of its height.
 
-*Data stack height* here refers to the number of stack values accessible by this function, i.e. it does not take into account values of caller functions' frames (but does include this function's inputs). Note that validation procedure does not require actual data stack implementation, but only to keep track of its height. Current height value starts at `type[code_section_index].inputs` (number of inputs of this function) at function entry and is updated at each instruction.
+*Terminating instructions* refers to the instructions either:
 
-At the same time the following properties are being verified:
+- ending function execution: `RETF`, or
+- ending whole EVM execution: `STOP`, `RETURN`, `REVERT`, `INVALID`.
```

Comment:
What is the expected stack height for these Opcode ? is it 0 ?

Reply:
There are no additional expectations besides the regular stack checks: e.g. for `RETURN` the stack height must be at least 2.